docs/guide/serialization.rst

```diff
@@ -0,0 +1,265 @@
+dictionary when it is deserializes (decodes) the data.
+
+Details of how the object is recognized for encoding or how the dictionary
+is recognized for decoding can be changed by passing functions to the
+``is_my_obj`` or ``is_my_dict`` parameters of :class:`.JSONCodec`.
+
+Dumping arbitrary objects to JSON
+---------------------------------
+
+Any :class:`.GufeTokenizable` can be dumped to a JSON file using the custom
+JSON handlers. Given a :class:`.GufeTokenizable` called ``obj`` and a
+path-like called ``filename``, you can dump to JSON with this recipe:
+
+.. code::
+
+    import json
+    from gufe.tokenization import JSON_HANDLER
+    with open(filename, mode='w') as f:
+        json.dump(obj.to_dict(), f, cls=JSON_HANDLER.encoder)
+
+Similarly, you can reload the object with:
+
+.. code::
+
+    import json
+    from gufe.tokenization import JSON_HANDLER
+    with open(filename, mode='r') as f:
+        obj = json.load(f, cls=JSON_HANDLER.decoder)
+
+Note that these objects are not space-efficient: that is, if you have
+the same object in memory stored at multiple locations (e.g., an identical
+``SolventComponent`` in more than one ``ChemicalSystem``), then you will
+save multiple copies of its JSON representation.
+
+.. Using JSON codecs outside of JSON
+.. ---------------------------------
+
+.. In a custom recursive storage scheme, 
+
+.. TODO: DWHS wants to write something here that describes how to use the
+   codecs in your own non-JSON storage scheme. But this is complicated
+   enough that it will take significant time
+
+When your object has recursive references
+-----------------------------------------
+
+In some cases, your object may have recursive references to other objects.
+For example, you may have objects ``parent`` and ``child``, where
+``parent.children`` includes ``child`` and ``child.parent`` is ``parent``.
+This means that the object dependency graph is not a directed acyclic graph.
+The best solution here is to avoid this design pattern whenever possible.
+Importantly, the ``child`` object in this case cannot be immutable, unless
+it is only created as a part of the creation of ``parent``.
+
+However, this could be made functionally immutable by (1) requiring that a
+valid ``child`` set its ``parent`` attribute exactly once; (2) not storing
+the ``child.parent`` attribute, and instead ensuring it gets set by the
+``parent`` (e.g., in ``__init__``). Note that this also assumes that all
+``children`` are provided to ``parent`` on initialization, otherwise
+``parent`` is also mutable.
+
+Another approach here would be to make it so that ``child`` was not a
+:class:`.GufeTokenizable`, and instead explicitly handle its serialization
+in a more complicated ``parent._to_dict`` method, with similarly more
+complicated ``parent._from_dict``. This approach is particularly useful if
+the ``child`` object isn't too complicated, and if the ``child`` is unlikely
+to be reused outside the context of the ``parent``.
```

Comment:
I added this because it was an issue we hit quickly once OPS had external contributors. People structured objects this way (it was also common to have sibling-sibling links) and so we needed to handle that serialization. I figured we might as well preempt problems here.

> Maybe a better paragraph would address how to structure parent child relationships (i.e. keep it one way, don't go circular)

I *do* say that the best solution is to avoid that design! But (at least in some OPS objects) it became impossible to completely avoid the design.

Reply:
@dotsdl: To clarify, the audience of this paragraph is the person who says "Why is my code giving me a `RecursionError` when I try to save my object to disk?"

We are not trying to do cycle detection, and we're not doing anything to allow the behavior. We're giving them 3 ways out:

1. Redesign your object to avoid this pattern
2. Ensure that your object is functionally immutable, and that the parent sets the child attributes
3. Remove the children from the gufe serialization.

Again, we hit this pretty quickly with outside contributions to OPS. OTOH, that might be a peculiarity of path sampling. In path sampling, it's pretty common to have two variants of a method (subclasses of an abstract) based on the direction of time. This has to do with maintaining detailed balance when performing MC of trajectories. If these variants need to share information or know about each other, then you end up with these circular references.

Maybe we won't hit that as quickly here, but since I've had to deal with it before, I'd already identified the ways to get around it. Maybe this can move elsewhere, but I think it's potentially useful information to provide.